### DISCUSSION TITLE
Terraform/OpenTofu Additional Run Task Options

### DISCUSSION THREAD

**Discussion by cindrmon** [semaphoreui/semaphore#2298] [Ideas]
Based from the Issue #2186 , i would want to know about the possible future implementations of improved Terraform/OpenTofu Task Run Checkboxes, as I have been eager to use OpenTofu with Semaphore UI, but only having the following options for TF/Tofu lackluster:

![current TF task options](https://github.com/user-attachments/assets/1368a45b-3634-4550-a331-dffb24875993)

i have a couple questions:

1. what is the default `terraform`/`tofu` command used in running a single Terraform/OpenTofu task? When I wanted to destroy the IaC I made, adding the `-destroy` tag doesn't seem to destroy the TFState? (see example below)

![use custom `-destroy` arg for destroying the current state](https://github.com/user-attachments/assets/9c0659e8-f2dc-4348-9296-b0db7846bb27)
![it always wants to create infrastructure even adding the argument](https://github.com/user-attachments/assets/46643d30-7cd4-45ab-b27e-41cf9bd34a6a)

2. is there any way for me to try to at least contribute an update adding the required checkboxes as provided by this comment?

From: https://github.com/semaphoreui/semaphore/issues/2186#issuecomment-2222583032
> I think we should have the following checkboxes:
> 
> * Plan
> * Apply
> * Destroy
> * Show

I really am interested in using this tool, and it has been a great help for me to have a self-hosted interface for both ansible and terraform/opentofu in my homelab, and i just want to know if there are plans on improving the usage for terraform/opentofu integration ever since it was first released a couple months ago? would want to know about updates with this project, and i would want to fully support it even if it means building out features myself if no one could. thanks

**Comment by DanatMighty** [comment]
I would really like to know how to destroy as well.

**Reply by fiftin** [reply]
Do you use Semaphore for Terraform?

**Reply by DanatMighty** [reply]
Yes using OpenTofu with Semaphore.

**Comment by damjank12** [comment]
Also here. We have huge number of artefacts being created for one time testing purposes, we gave teams access to click and play. But now we are facing that huge number of for instance EKS cllusters are bing created, tested and still be running since there is no working way for the user to issue destroy command, once done. Any info how to go about that? Thank you for superb tool so far!

**Reply by fiftin** [reply]
Do you use Semaphore for Terraform?

**Comment by madozu** [comment]
+1 for adding a "Destroy" checkbox ... that would really be helpful 👍

**Reply by fiftin** [reply]
Hi @madozu ,

do you use Terraform/OpenTofu in Semaphore? I need your feedback ;)

**Reply by madozu** [reply]
Hi @fiftin ,

yes I am using Terraform in Semaphore Version 2.10.22 to deploy some test VMs / Containers. I can also provide testing / feedback. For now I am just deleting them manually when they're not needed anymore.

**Comment by mlanner** [comment]
@fiftin Here's another :+1: ... I just started using Semaphore (v2.10.21) + Terraform/OpenTofu and I too noticed the lack of a **destroy** option. Just like others in this thread I would really like to have that as well. I do understand that there might be reasons around privileges that made you not include it. That said, I think if someone is an **owner** or **manager** on the project, they should be able to destroy an environment. For members who are just **task runner** or below, they probably shouldn't be able to destroy resources. Well, at least now without permission by someone with a higher privilege. In my use-case, most, if not all, members would be at a minimum managers and would be trusted to do whatever they deem necessary, but I realize that might not be the case in all situations.

I'd be happy to provide more input and feedback if needed. Thanks for a great project/product!

**Comment by allancassaro** [comment]
I know this isn't optimal, but for those like me that are trying to destroy resources with Semaphore and Terraform/OpenTofu, it is possible to add this:
```
TF_CLI_ARGS_apply="-destroy=true"
```
As an environment variable.
Just don't forget to remove the variable after, otherwise it will try to destroy on every run...

It is hacky, ugly, but it works!

More about supported environment variables here: https://developer.hashicorp.com/terraform/cli/config/environment-variables

**Comment by madozu** [comment]
With version v2.11.0-beta3 the "destroy" option is implemented along with a fix for the "auto-approve" option requiring confirmation.

Works great for me 👍, thanks a lot for the team implementing this 🥇
